ldclient_scoped.go

```diff
@@ -0,0 +1,557 @@
+// operations on it, like flag evaluations and event tracking.
+//
+// All of the contexts added must have kinds that are not already present. If you
+// try to add a context with a duplicate kind, the new context will not be added
+// and a warning will be logged. If you want to replace an existing context, use
+// [LDScopedClient.OverwriteContextByKind] instead.
+//
+// The scoped client's contexts so far are combined into a multi-context whenever the
+// scoped client is used.
+func (c *LDScopedClient) AddContext(contexts ...ldcontext.Context) {
+	c.Lock()
+	defer c.Unlock()
+	c.rebuild = true
+
+	for _, ctx := range contexts {
+		if ctx.Multiple() {
+			for _, individual := range ctx.GetAllIndividualContexts(nil) {
+				c.addIndividualContext(individual)
+			}
+			continue
+		}
+		c.addIndividualContext(ctx)
+	}
+}
+
+func (c *LDScopedClient) overwriteIndividualContextByKind(context ldcontext.Context) {
+	c.contexts[context.Kind()] = context
+}
+
+// OverwriteContextByKind overwrites an existing context in the scoped client, affecting all future
+// operations on it, like flag evaluations and event tracking.
+//
+// You may pass a multi-context, or multiple individual contexts, to OverwriteContextByKind.
+// Any existing contexts with the same kind as any of the new contexts will be overwritten.
+//
+// If the scoped client had multiple contexts, only the context with the same kind as the new context
+// will be overwritten. Any other existing contexts will remain unchanged:
+//
+//	company := ldcontext.NewWithKind("company", "acme")
+//	scopedClient := client.ForContext(company)
+//	scopedClient.AddContext(ldcontext.New("user-key"))
+//	scopedClient.OverwriteContextByKind(ldcontext.NewWithKind("company", "monsoon"))
+//	scopedClient.CurrentContext() // returns a multi-context with "monsoon" and "user-key"
+func (c *LDScopedClient) OverwriteContextByKind(contexts ...ldcontext.Context) {
+	c.Lock()
+	defer c.Unlock()
+	c.rebuild = true
+
+	for _, ctx := range contexts {
+		if ctx.Multiple() {
+			for _, individual := range ctx.GetAllIndividualContexts(nil) {
+				c.overwriteIndividualContextByKind(individual)
+			}
+			continue
+		}
+		c.overwriteIndividualContextByKind(ctx)
+	}
+}
+
+// CurrentContext returns the current context for the scoped client. This is the
+// combination of all the contexts that have been added so far, as a
+// multi-context.
+func (c *LDScopedClient) CurrentContext() ldcontext.Context {
+	c.Lock()
+	defer c.Unlock()
+	if !c.rebuild {
+		return c.context
+	}
+	c.rebuild = false
+	b := ldcontext.NewMultiBuilder()
+	for _, ctx := range c.contexts {
+		b.Add(ctx)
+	}
+	c.context = b.Build()
+	return c.context
+}
+
+// Client returns the underlying LDClient that this scoped client wraps.
+func (c *LDScopedClient) Client() *LDClient {
+	return c.client
+}
+
+// Contextual methods: equivalent to calling the same method on the underlying client with the current context
```

Comment:
All of the remaining methods in this file mirror some existing method on `LDClient`, but instead of taking a `context`, use the `c.CurrentContext()`.

All the doc comments are copied over - I figured this would be better than just `See LDClient.MethodName for more information`, because once a customer adopts the scoped client, there will be developers in that organization that _only_ ever interact with the scoped client, and they may need quick explainers of each method if it's their first time using LD clients at all.